File: FamHelpDeskBackend/models/notification_settings.py
from models.base import FamHelpDeskBaseModel
from pynamodb.attributes import UnicodeAttribute, BooleanAttribute, NumberAttribute


class NotificationSettingsModel(FamHelpDeskBaseModel):
    """
    Model for storing user notification preferences.

    PK: USER_PROFILE#{user_id}
    SK: NOTIFICATION_SETTINGS
    """

    user_id = UnicodeAttribute()

    # Notification type preferences with defaults
    welcome_enabled = BooleanAttribute(default=True)
    membership_enabled = BooleanAttribute(default=True)
    ticket_creation_enabled = BooleanAttribute(default=True)
    ticket_assigned_enabled = BooleanAttribute(default=True)
    ticket_comment_enabled = BooleanAttribute(default=False)
    ticket_status_changed_enabled = BooleanAttribute(default=False)
    group_invitation_enabled = BooleanAttribute(default=False)

    # Standard timestamp fields
    created_date = NumberAttribute()
    last_updated = NumberAttribute()

    @staticmethod
    def create_pk(user_id: str) -> str:
        """Create partition key for notification settings"""
        return f"USER_PROFILE#{user_id}"

    @staticmethod
    def create_sk() -> str:
        """Create sort key for notification settings"""
        return "NOTIFICATION_SETTINGS"

    @staticmethod
    def clean_returned_settings(settings: "NotificationSettingsModel") -> dict:
        """Clean and return notification settings data"""
        return {
            "user_id": settings.user_id,
            "welcome_enabled": settings.welcome_enabled,
            "membership_enabled": settings.membership_enabled,
            "ticket_creation_enabled": settings.ticket_creation_enabled,
            "ticket_assigned_enabled": settings.ticket_assigned_enabled,
            "ticket_comment_enabled": settings.ticket_comment_enabled,
            "ticket_status_changed_enabled": settings.ticket_status_changed_enabled,
            "group_invitation_enabled": settings.group_invitation_enabled,
            "created_date": settings.created_date,
            "last_updated": settings.last_updated,
        }

File: FamHelpDeskiOS/FamHelpDesk/FamHelpDesk/Networking/Models/NotificationSettings.swift
import Foundation

struct NotificationSettingsResponse: Codable {
    let settings: NotificationSettings
}

struct NotificationSettings: Codable {
    let userId: String
    let welcomeEnabled: Bool
    let membershipEnabled: Bool
    let ticketCreationEnabled: Bool
    let ticketAssignedEnabled: Bool
    let ticketCommentEnabled: Bool
    let ticketStatusChangedEnabled: Bool
    let groupInvitationEnabled: Bool
    let createdDate: Int
    let lastUpdated: Int

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case welcomeEnabled = "welcome_enabled"
        case membershipEnabled = "membership_enabled"
        case ticketCreationEnabled = "ticket_creation_enabled"
        case ticketAssignedEnabled = "ticket_assigned_enabled"
        case ticketCommentEnabled = "ticket_comment_enabled"
        case ticketStatusChangedEnabled = "ticket_status_changed_enabled"
        case groupInvitationEnabled = "group_invitation_enabled"
        case createdDate = "created_date"
        case lastUpdated = "last_updated"
    }
}

struct UpdateNotificationSettingsRequest: Codable {
    let welcomeEnabled: Bool?
    let membershipEnabled: Bool?
    let ticketCreationEnabled: Bool?
    let ticketAssignedEnabled: Bool?
    let ticketCommentEnabled: Bool?
    let ticketStatusChangedEnabled: Bool?
    let groupInvitationEnabled: Bool?

    enum CodingKeys: String, CodingKey {
        case welcomeEnabled = "welcome_enabled"
        case membershipEnabled = "membership_enabled"
        case ticketCreationEnabled = "ticket_creation_enabled"
        case ticketAssignedEnabled = "ticket_assigned_enabled"
        case ticketCommentEnabled = "ticket_comment_enabled"
        case ticketStatusChangedEnabled = "ticket_status_changed_enabled"
        case groupInvitationEnabled = "group_invitation_enabled"
    }
}
